pkg/util/cache/basic_cache.go

```diff
@@ -27,7 +38,7 @@ func (b *BasicCache) Add(k string, v interface{}) bool {
 	defer b.m.Unlock()
 
 	current, found := b.cache[k]
-	if !found || current != v {
+	if !found || !b.eqFunc(current, v) {
```

Comment:
Is this function call often as using reflect may have an impact on performance?

Reply:
The only place where BasicCache is used is `pkg/collector/corechecks/embed/jmx/state.go`, switching to deepEqual was mostly because we allow virtually anything to be stored inside the cache, including uncomparable type (that led to the crash this PR aim to fix). So effectively at runtime `reflect.deepEqual` is never ever used because in the only place we use the cache uses a custom equality function.

Reply:
Was this logic of comparing the configs stored under the same key ever useful?

Given that this would crash before, I tend to think the comparison logic was never useful.
While the Agent loads the initial integration configs, if a config is identical to one that's already stored, we could simply overwrite the old one with the new one and update the `modified` timestamp.
When the AD later schedules configs, I believe it is responsible for not scheduling configs that are identical to configs that are already scheduled. So this logic here seems to add useless complexity.

WDYT? (please check the assertions I made in this comment on how other components should behave :) )

Reply:
I doubled check the autodiscovery logic, I'm not 100% we can't get duplicated check. However the agent will never schedule twice a check with the same Digest (as per `Collector.RunCheck`). So, as of today, outside of `agent jmx collect` / `agent check <jmx_check>` the comparison was never used during agent life as the addition the to cache is happening subsequently through `JMXCheck.Run` (which make sense else we would probably noticed that this same config situation crashed the agent).

The timestamp being rather useless for `agent jmx collect` / `agent check <jmx_check>` just removing the comparison will indeed lead to the exact same behaviour for `agent start` while it won't change much for `agent jmx collect` & `agent check <jmx_check>`. I would say it's safe to just remove the comparison.

Reply:
thanks for checking, let's remove the comparison logic altogether then 👍 

Btw, `git blame` tells me this comparison logic was added in https://github.com/DataDog/datadog-agent/pull/4228 so that the `BasicCache` could be used by the inventories logic, but looks like it didn't end up being used for inventories. In any case, makes sense to remove this logic now.